### ISSUE TITLE
Error in compiling a C program for ALPHA: " no .eh_frame_hdr table will be created."

### ISSUE THREAD

**Issue by jerrysinha** [crosstool-ng/crosstool-ng#719]
I have build the cross compiler for ALPHA using crosstool-ng. I am not able to compile any C/C++ code. It is throwing the following error:

:~/Desktop$ alphaev4-unknown-linux-gnu-g++ -static new1.cpp -o hello
/home/iiitd/local/x-tools/alphaev4-unknown-linux-gnu/lib/gcc/alphaev4-unknown-linux-gnu/6.3.1/../../../../alphaev4-unknown-linux-gnu/bin/ld: error in /home/iiitd/local/x-tools/alphaev4-unknown-linux-gnu/lib/gcc/alphaev4-unknown-linux-gnu/6.3.1/crtend.o(.eh_frame); no .eh_frame_hdr table will be created.

- crosstool-NG version: crosstool-ng-1.23.0.rc2
- Host OS: ubuntu 16.04 LTS, 64-bit

**Comment by stilor** [comment]
.config?

**Comment by jerrysinha** [comment]
I am new to crosstool-ng. I have searched for the .config file you are asking for, but cannot locate it. Please guide me from where I can get the .config file.
I have the build.log file. I am attaching it.
https://drive.google.com/file/d/0B3MslpubFyURMFdlNE85Mi1hQmM/view?usp=sharing

**Comment by jerrysinha** [comment]
Hi @stilor I am attaching the .config file...please kindly find it in the link.
https://drive.google.com/file/d/0B3MslpubFyURM21iNjkxOU5wQms/view?usp=sharing

**Comment by stilor** [comment]
I won't be able to look at it till Sunday. And even then, don't get your hopes high - it is likely not an issue with crosstools, most likely it is a regression in binutils. I'll try to pinpoint when it started - but my knowledge on binutils externals may not be sufficient to fix it.

**Comment by jerrysinha** [comment]
Okay @stilor . I came across this link "https://github.com/crosstool-ng/crosstool-ng/issues/370" which had the same issue.. According to the issue, I tried to build with crosstool-ng-1.15.0. The steps are as follows:

$ mkdir -p ~/src/gem2/toolchain  
$ cd ~/src/gem2/toolchain  
$  wget http://crosstool-ng.org/download/crosstool-ng/crosstool-ng-1.15.0.tar.bz2  
$ tar xjf crosstool-ng-1.15.0.tar.bz2  
$ cd crosstool-ng-1.15.0/  
$ ./configure -prefix=/home/iiitd/local/gem2-crosstool-ng  
--------------------------------
---------------------------------
configure: error: could not find GNU libtool >= 1.5.26
$ sudo apt-get install libtool-bin
$ ./configure -prefix=/home/iiitd/local/gem2-crosstool-ng
$ make
**Makefile:114: *** Recursion detected, bailing out....  Stop.
Makefile:119: recipe for target 'build' failed
make: *** [build] Error 2**

I am attaching the config.log.  
https://drive.google.com/file/d/0B3MslpubFyURMTBjcjhFZlJXeWM/view?usp=sharing

**Comment by stilor** [comment]
This one is easy to fix (just add the two extra lines checking for MAKEFLAGS which are in the Makefile.in in master). But there are other issues, e.g. gcc-4.3 (that is the default in that sample) cannot be built with gcc6 (which is the default on modern distributions).

Thanks for pointing that issue, I'll try to use it as a baseline.

**Comment by jerrysinha** [comment]
Thank you very much @stilor for your response.
But I have not understood which lines you are indicating. I apologize beforehand if I am asking a silly question. Are you suggesting me to add the below lines in the makefile?

    38  # Helper: print abbreviation of the command by default, or full command
    39 # if doing 'make V=1'.
    40  __silent = $(if $(V),,@printf '  %-7s %s\n' '$1' '$(if $2,$2,$(strip $<))' && )
    41  __silent_rm = $(call __silent,RM,$1)rm -f $1
    42  __silent_rmdir = $(call __silent,RMDIR,$1)rm -rf $1

**Comment by stilor** [comment]
````
ifeq ($(filter Rr,$(MAKEFLAGS)),)
ifeq ($(filter -Rr,$(MAKEFLAGS)),)
ifeq ($(filter rR,$(MAKEFLAGS)),)
ifeq ($(filter -rR,$(MAKEFLAGS)),)
CT_MAKEFLAGS += -Rr
endif # No -rR
endif # No rR
endif # No -Rr
endif # No Rr
````
https://github.com/crosstool-ng/crosstool-ng/blob/master/Makefile.in#L28

**Comment by stilor** [comment]
So, first thing is that this issue is not an *error* - GCC still creates a binary. Also, .eh_frame_hdr section is essentially an index into the .eh_frame section (which *is* created), so the exception handling may even work with the resulting binaries (albeit much slower, since it would have to resort to linear search rather than binary search).

I bisected the issue yesterday:
- GCC 4.7.x and earlier do not have this issue
- GCC 4.8.x have this issue starting with 4.8.4 (4.8.3 is okay)
- GCC 4.9.x and later all have this issue

The issue started with this commit: https://github.com/gcc-mirror/gcc/commit/66035fd81f6fb8dff84e0c64d52ed041450fdebc; the problem it was trying to solve is described here: https://gcc.gnu.org/ml/gcc-patches/2014-07/msg01680.html

However, it is a red herring: this only manifested the issue. The problem is that the crtend.o produced by as has these DWARF records:

````
~/local-bad/x-tools/alphaev4-unknown-linux-gnu/bin/alphaev4-unknown-linux-gnu-readelf -wf .build/alphaev4-unknown-linux-gnu/buildtools/lib/gcc/alphaev4-unknown-linux-gnu/4.8.4/crtend.o
Contents of the .eh_frame section:

00000000 ZERO terminator


00000004 0000000000000010 00000000 CIE
  Version:               1
  Augmentation:          "zR"
  Code alignment factor: 4
  Data alignment factor: -8
  Return address column: 26
  Augmentation data:     1b

  DW_CFA_def_cfa_register: r30
  DW_CFA_nop

00000018 0000000000000024 00000018 FDE cie=00000004 pc=0000000000000000..0000000000000070
  DW_CFA_advance_loc: 20 to 0000000000000014
  DW_CFA_def_cfa_offset: 32
  DW_CFA_advance_loc: 12 to 0000000000000020
  DW_CFA_offset: r9 at cfa-24
  DW_CFA_advance_loc: 8 to 0000000000000028
  DW_CFA_offset: r10 at cfa-16
  DW_CFA_advance_loc: 8 to 0000000000000030
  DW_CFA_offset: r26 at cfa-32
  DW_CFA_advance_loc: 60 to 000000000000006c
  DW_CFA_restore: r10
  DW_CFA_restore: r9
  DW_CFA_restore: r26
  DW_CFA_def_cfa_offset: 0
  DW_CFA_nop
  DW_CFA_nop
  DW_CFA_nop
  DW_CFA_nop
  DW_CFA_nop
````
which have the `ZERO terminator` as the first record. This trips the BFD reader of the EH frame info: https://sourceware.org/git/gitweb.cgi?p=binutils-gdb.git;a=blob;f=bfd/elf-eh-frame.c;h=52ba9c62138bb7d2c8901d961ba322dbfe23e220;hb=HEAD#l724:
````
 724      if (hdr_length == 0)
 725         {
 726           /* A zero-length CIE should only be found at the end of
 727              the section.  */
 728           REQUIRE ((bfd_size_type) (buf - ehbuf) == sec->size);
 729           ENSURE_NO_RELOCS (buf);
 730           sec_info->count++;
 731           break;
 732         }
````
and produces this error. I still need to find out why `gas` has generated such zero-sized CIE first.

**Comment by jerrysinha** [comment]
Thanks for your response and suggestions..

**Comment by luizfernandonb** [comment]
This has already been "fixed" since GCC 7, however I'm using GCC 9.5.0 and I'm still having this problem, I searched the source code and this patch is already applied, but the same warning still appears when I compile some C file
